### ISSUE TITLE
Ending a state definition

### ISSUE THREAD

**Issue by Nalha** [pact-foundation/pact-jvm#497]
In pact-jvm-consumer-junit (and others?) there is currently no way to end a "given" definition, this makes it easy to accidentally place more requests that should'nt be in that state.
One way around this is using states for all requests but I don't think that should feel as forced as it currently is.

```
.given("mystate")
    .uponReceiving("post my request")
        .matchPath("/api/.+/.+/.+")
        .method("POST")
    .willRespondWith()
        .status(200)
        .body("string result")`
.endGiven()                       <---
    .uponReceiving...
````

**Comment by uglyog** [comment]
The state only applies to the proceeding interaction. The next `.uponReceiving` call resets the state to none if there was not another `.given` that preceded it. 

So

```java
    .given("mystate")
    .uponReceiving("post my request") // This interaction has state 'mystate'
        .matchPath("/api/.+/.+/.+")
        .method("POST")
    .willRespondWith()
        .status(200)
        .body("string result")`
    .uponReceiving... // This interaction will have no state, as it resets the state and there was no preceeding .given call
```

**Comment by Nalha** [comment]
That might be the idea but it is not how it currently works, in consumer-junit 3.5.2 all 3 uponReceiving following my test state got the following in the JSON:

```
            "providerStates": [
                {
                    "name": "test state"
                }
            ]
```

**Comment by uglyog** [comment]
That's a defect then.

**Comment by solarmosaic-kflorence** [comment]
@uglyog I just saw this as well, with version `4.2.8` of `pact-jvm`:

```
  ConsumerPactBuilder
    .consumer(consumerName)
    .hasPactWith(providerName)
    .given("greeting", Map("name" -> "world").asJava)
    .uponReceiving("GET /hello")
    .path("/hello")
    .method("POST")
    .willRespondWith()
    .status(200)
    .body(Json.toJson(HelloResponse("world")).toString())
    .uponReceiving("GET /hello-user")
    .path("/hello-user")
    .method("POST")
    .willRespondWith()
    .status(200)
    .body(Json.toJson(HelloResponse("user")).toString())
    .toPact()
```

Results in the following pact file:

```
{
  "consumer": {
    "name": "consumer"
  },
  "interactions": [
    {
      "description": "GET /hello",
      "providerStates": [
        {
          "name": "greeting",
          "params": {
            "name": "world"
          }
        }
      ],
      "request": {
        "method": "POST",
        "path": "/hello"
      },
      "response": {
        "body": {
          "hello": "world"
        },
        "status": 200
      }
    },
    {
      "description": "GET /hello-user",
      "providerStates": [
        {
          "name": "greeting",
          "params": {
            "name": "world"
          }
        }
      ],
      "request": {
        "method": "POST",
        "path": "/hello-user"
      },
      "response": {
        "body": {
          "hello": "user"
        },
        "status": 200
      }
    }
  ],
  "metadata": {
    "pact-jvm": {
      "version": "4.2.8"
    },
    "pactSpecification": {
      "version": "3.0.0"
    }
  },
  "provider": {
    "name": "example"
  }
}
```

**Comment by solarmosaic-kflorence** [comment]
The problem seems to be here: https://github.com/pact-foundation/pact-jvm/blob/c77faa70db24c745b4e1078ef454192cf3213a3c/consumer/src/main/kotlin/au/com/dius/pact/consumer/dsl/PactDslResponse.kt#L339

It is starting a new interaction using the previous `PactDslRequestWithPath`, which carries over the `consumer`, `provider`, `state` and `path`.

**Comment by uglyog** [comment]
@solarmosaic-kflorence good catch!

**Comment by uglyog** [comment]
Released 4.1.24 and 4.2.9

**Comment by solarmosaic-kflorence** [comment]
@uglyog confirmed:

```json
{
  "consumer": {
    "name": "consumer"
  },
  "interactions": [
    {
      "description": "GET /hello-user",
      "request": {
        "method": "POST",
        "path": "/hello-user"
      },
      "response": {
        "body": {
          "hello": "user"
        },
        "status": 200
      }
    },
    {
      "description": "GET /hello",
      "providerStates": [
        {
          "name": "greeting",
          "params": {
            "name": "world"
          }
        }
      ],
      "request": {
        "method": "POST",
        "path": "/hello"
      },
      "response": {
        "body": {
          "hello": "world"
        },
        "status": 200
      }
    }
  ],
  "metadata": {
    "pact-jvm": {
      "version": "4.2.9"
    },
    "pactSpecification": {
      "version": "3.0.0"
    }
  },
  "provider": {
    "name": "example"
  }
}
```

I think this ticket can be closed.
